File: frontend/src/app/features/pedidos/domain/interfaces/repositories/IProductoRepository.ts
import type { Producto } from '../../entities/producto';

/**
 * Contrato del repositorio de productos.
 * @remarks Implementado por {@link ProductoRepository}.
 */
export interface IProductoRepository {
  /** Devuelve todos los productos del catálogo. */
  getAll(): Promise<Producto[]>;
  /** Devuelve únicamente los productos activos. */
  getActivos(): Promise<Producto[]>;
  /**
   * Obtiene un producto por su identificador.
   * @param id Identificador del producto.
   * @returns Entidad {@link Producto} o `null` si no existe.
   */
  getById(id: number): Promise<Producto | null>;
  /**
   * Filtra productos por categoría.
   * @param categoriaId Identificador de la categoría.
   */
  getByCategoria(categoriaId: number): Promise<Producto[]>;
}

File: backend/ERP.Backend/Domain/Interfaces/IRepository/IProductoRepository.cs
using Domain.Entities;

namespace Domain.Interfaces
{
    /// <summary>
    /// Interfaz del repositorio del catálogo de productos.
    /// </summary>
    public interface IProductoRepository
    {
        /// <summary>
        /// Obtiene todos los productos del catálogo con su categoría asociada.
        /// </summary>
        /// <returns>Lista de <see cref="Producto"/> con la entidad <see cref="CategoriaProducto"/> cargada.</returns>
        Task<List<Producto>> GetAllAsync();

        /// <summary>
        /// Obtiene un producto por su identificador único con su categoría asociada.
        /// </summary>
        /// <param name="productoID">Identificador único del producto a buscar.</param>
        /// <returns>El <see cref="Producto"/> encontrado con su categoría cargada, o <c>null</c> si no existe.</returns>
        Task<Producto?> GetByIdAsync(int productoID);

        /// <summary>
        /// Obtiene todos los productos pertenecientes a una categoría específica.
        /// </summary>
        /// <param name="categoriaID">Identificador de la categoría por la que filtrar.</param>
        /// <returns>Lista de <see cref="Producto"/> de la categoría indicada con su categoría cargada.</returns>
        Task<List<Producto>> GetByCategoriaAsync(int categoriaID);
    }
}